### PR TITLE
[Refactor] Extract workload classes for reduction ops

### LINKED ISSUE #825: [REFACTOR][Reduction] Extract workload classes and align test imports with trust model
### Symptom / Motivation

The reduction family migration (PR #794, #798, #812, #818) updated all 19 ops to spec-conformant interfaces, but the supporting test infrastructure was not refactored to match the trust model architecture defined in `docs/testing.md` and `docs/trust-model.md`.

Currently, 4 reduction test files define `gen_inputs()` inline (inside `TestBase` subclasses) instead of importing from `workloads/ops/`. The workloads layer files do not exist for these op groups. By contrast, the softmax family was fully aligned during its migration (PR #794): `workloads/ops/softmax.py` exists and `tests/ops/test_softmax.py` imports from it.

This inconsistency blocks the benchmark migration (follow-up issue) because benchmark files need to import workload classes from `workloads/ops/` without importing test logic — a trust model boundary.

### Root Cause Analysis

Migration inconsistency. PR #787 (workloads layer detach) established the pattern, and PR #794 (softmax) followed it correctly. PR #798 (remaining 17 ops) only updated Op constructors without restructuring the test/workload layers. The 4 test files predate the trust model architecture.

### Related Files

- `tests/ops/test_reduce.py` — `ReduceTest(TestBase)` with inline `gen_inputs()`
- `tests/ops/test_argreduce.py` — inline `gen_inputs()`
- `tests/ops/test_logical_reduce.py` — inline `gen_inputs()`
- `tests/ops/test_vector_norm.py` — inline `gen_inputs()`
- `workloads/ops/softmax.py` — reference implementation (correct pattern)
- `tests/ops/test_softmax.py` — reference (imports from `workloads/ops/softmax`)
- `docs/testing.md` — file checklist requiring workload class in `workloads/ops/`
- `docs/trust-model.md` — workloads layer ownership and boundary rules

## Goal

Create `workloads/ops/` files for the 4 reduction op groups and refactor test files to import from them, aligning with the trust model's workloads layer contract.

## Plan

1. Create `workloads/ops/reduce.py` with `WorkloadBase` subclasses for sum/mean/amax/amin/prod/std/var/var_mean (`gen_inputs()` only, no `ref_program`)
2. Create `workloads/ops/argreduce.py` with workload classes for argmax/argmin
3. Create `workloads/ops/logical_reduce.py` with workload classes for all/any/count_nonzero
4. Create `workloads/ops/vector_norm.py` with workload classes for l1_norm/l2_norm/inf_norm
5. Refactor `tests/ops/test_reduce.py`: change `ReduceTest` and `WelfordTest` to inherit from workload classes (`class ReduceTest(ReduceWorkload, TestBase)`), keep `ref_program()` in test
6. Refactor `tests/ops/test_argreduce.py`, `test_logical_reduce.py`, `test_vector_norm.py` similarly
7. Verify all existing tests pass with no behavioral change

## Constraints

- MUST NOT modify `benchmarks/`, `tileops/ops/`, or `tileops/ops_manifest.yaml`
- Workload files MUST contain only `gen_inputs()` — no `ref_program`, no `check()`, no `calculate_flops/memory`, no benchmark baselines (trust-model.md §Workloads Layer)
- Existing test behavior and parametrization must be preserved — this is a pure structural refactor
- Follow the pattern established by `workloads/ops/softmax.py` and `tests/ops/test_softmax.py`

## Acceptance Criteria

- [ ] AC-1: Modified files pass unit tests (`pytest tests/ops/test_{reduce,argreduce,logical_reduce,vector_norm}.py`)
- [ ] AC-2: `workloads/ops/{reduce,argreduce,logical_reduce,vector_norm}.py` exist, each exporting only `WorkloadBase` subclasses with `gen_inputs()`
- [ ] AC-3: Test files import workload classes from `workloads/ops/`, not defining `gen_inputs()` inline
- [ ] AC-4: No `ref_program` or test-layer logic in `workloads/ops/` files

### PR DESCRIPTION
**Pull request by lcy-seso**
## Summary

Create `workloads/ops/` files for the 4 reduction op groups (reduce, argreduce, logical_reduce, vector_norm) and refactor test files to import from them, aligning with the trust model's workloads layer contract.

Closes #825

## Test plan

- [x] **AC-1**: Modified files pass unit tests (`pytest tests/ops/test_{reduce,argreduce,logical_reduce,vector_norm}.py`)
- [x] **AC-2**: `workloads/ops/{reduce,argreduce,logical_reduce,vector_norm}.py` exist, each exporting only `WorkloadBase` subclasses with `gen_inputs()`
- [x] **AC-3**: Test files import workload classes from `workloads/ops/`, not defining `gen_inputs()` inline
- [x] **AC-4**: No `ref_program` or test-layer logic in `workloads/ops/` files

Tests: 413/413 passed, 0 failed

## Changes

- New `workloads/ops/reduce.py` — workload classes for sum/mean/amax/amin/prod/std/var/var_mean
- New `workloads/ops/logical_reduce.py` — workload classes for all/any/count_nonzero
- New `workloads/ops/vector_norm.py` — workload classes for l1_norm/l2_norm/inf_norm
- Refactored `tests/ops/test_reduce.py`, `test_argreduce.py`, `test_logical_reduce.py`, `test_vector_norm.py` to inherit from workload classes

## Follow-up

- #897 — Extract `_LogicalTest` base class in `logical_reduce.py` to deduplicate identical `__init__`/`gen_inputs` across `AnyTest`, `AllTest`, `CountNonzeroTest`

Suggestions: consolidate cross-module `_RandnTest` duplication; fix `_make_logical_input` docstring (says "2-D+" but applies to 1-D too)

### CODE REVIEW

File: workloads/ops/reduce.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+class VarMeanWorkload(WorkloadBase):
+    """Workload definition for VarMeanFwdOp."""
+
+    def __init__(self, shape: tuple, dtype: torch.dtype):
+        self.shape = shape
+        self.dtype = dtype
+
+    def gen_inputs(self) -> tuple[torch.Tensor]:
+        x = torch.randn(*self.shape, dtype=self.dtype, device="cuda")
+        return (x,)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply workloads/ops/reduce.py]: Adopted. Extracted `_RandnWorkload` base class in reduce.py, vector_norm.py, and argreduce.py. See 6e56822.

File: tests/ops/test_argreduce.py [hunk 2]
```diff
@@ -9,6 +9,7 @@
 import torch
 
 from tests.test_base import FixtureBase, TestBase
+from workloads.ops.argreduce import ArgmaxTest as _ArgmaxWorkload
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_argreduce.py:12]: Adopted. Renamed to `ArgmaxWorkload`/`ArgminWorkload` with backward-compatible aliases for existing benchmark imports. See 6e56822.

File: workloads/ops/logical_reduce.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+# ---------------------------------------------------------------------------
+
+
+def _make_logical_input(shape: tuple, dtype: torch.dtype) -> torch.Tensor:
+    """Create a tensor with a mix of zeros and non-zeros.
+
+    For 2-D+ inputs the first row is forced to all-zero (meaningful for
+    ``any``) and the second row to all-nonzero (meaningful for ``all``).
+    """
+    m = shape[0] if len(shape) >= 1 else 1
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply workloads/ops/logical_reduce.py:64]: Already handled in current code at 6e56822.

Evidence:
- `workloads/ops/logical_reduce.py:63` reads: `m = shape[0] if len(shape) >= 1 else 1`
- This is functionally equivalent to the suggested `m = shape[0] if shape else 1` -- both guard against empty-tuple shapes by defaulting `m` to 1.
- The 0-D tensor concern raised in this review is already addressed.

Current HEAD: 6e56822

File: workloads/ops/reduce.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply workloads/ops/reduce.py]: Addressed in 6e56822.

Evidence:
- Extracted `_RandnWorkload` base class in `workloads/ops/reduce.py` (line 6-15)
- `SumWorkload`, `MeanWorkload`, `AmaxWorkload`, `AminWorkload`, `StdWorkload`, `VarWorkload`, `VarMeanWorkload` now inherit from `_RandnWorkload` with no duplicated `__init__`/`gen_inputs`
- Same pattern applied in `workloads/ops/argreduce.py` and `workloads/ops/vector_norm.py`

Current HEAD: 6e56822

File: tests/ops/test_argreduce.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_argreduce.py:12]: Addressed in 6e56822.

Evidence:
- `ArgmaxTest` renamed to `ArgmaxWorkload` in `workloads/ops/argreduce.py` (line 18)
- `ArgminTest` renamed to `ArgminWorkload` (line 22)
- Backward-compatible aliases retained (lines 27-28)
- `tests/ops/test_argreduce.py` now imports `ArgmaxWorkload` (line 12)

Current HEAD: 6e56822

File: tests/ops/test_reduce.py [hunk 4]
```diff
@@ -97,29 +102,18 @@ class BesselFixture(FixtureBase):
...
 
 # ---------------------------------------------------------------------------
-# TestBase helpers
+# TestBase helpers — inherit gen_inputs() from workload classes
 # ---------------------------------------------------------------------------
 
 
-class ReduceTest(TestBase):
-    """Parameterized test helper for simple reduce ops."""
+class ReduceTest(SumWorkload, TestBase):
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_reduce.py]: This is by design at 6e56822.

Evidence:
- `ReduceTest` inherits from `SumWorkload` which inherits from `_RandnWorkload` (the common base added in 6e56822)
- `_RandnWorkload` is intentionally private (`_`-prefixed) and not exported from `workloads/ops/reduce.py` -- it's an implementation detail, not part of the workloads API
- All `_RandnWorkload` subclasses (`SumWorkload`, `MeanWorkload`, etc.) provide identical `gen_inputs` behavior via inheritance, so which one is chosen as the MRO parent is functionally irrelevant
- Exposing `_RandnWorkload` as a public API would break the workloads-layer contract: each public class maps 1:1 to an op

Current HEAD: 6e56822

File: tests/ops/test_argreduce.py [hunk 5]
```diff
@@ -101,23 +102,17 @@ class SpecArgreduceFixture(FixtureBase):
 
 
 # ---------------------------------------------------------------------------
-# TestBase helpers
+# TestBase helpers — inherit gen_inputs() from workload classes
 # ---------------------------------------------------------------------------
 
 
-class ArgreduceTest(TestBase):
+class ArgreduceTest(_ArgmaxWorkload, TestBase):
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_argreduce.py:109]: This is by design at 6e56822.

Evidence:
- `ArgreduceTest` inherits from `_ArgmaxWorkload` (locally aliased from `ArgmaxWorkload`)
- `ArgmaxWorkload` inherits from `_RandnWorkload` (added in 6e56822), same as `ArgminWorkload`
- Both argreduce workload classes share identical `gen_inputs` via `_RandnWorkload`, so the choice of parent is functionally equivalent
- `_RandnWorkload` is private and intentionally not exported -- the workloads API exposes one class per op

Current HEAD: 6e56822

File: workloads/ops/reduce.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+class MeanWorkload(_RandnWorkload):
+    """Workload definition for MeanFwdOp."""
+
+
+class AmaxWorkload(_RandnWorkload):
+    """Workload definition for AmaxFwdOp."""
+
+
+class AminWorkload(_RandnWorkload):
+    """Workload definition for AminFwdOp."""
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply workloads/ops/reduce.py]: Adopted. Renamed all `*Workload` classes back to `*Test` to match codebase convention (50+ existing classes). See a892135.

File: workloads/ops/vector_norm.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+class L1NormWorkload(_RandnWorkload):
+    """Workload definition for L1NormFwdOp."""
+
+
+class L2NormWorkload(_RandnWorkload):
+    """Workload definition for L2NormFwdOp."""
+
+
+class InfNormWorkload(_RandnWorkload):
+    """Workload definition for InfNormFwdOp."""
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply workloads/ops/vector_norm.py]: Adopted. Renamed all `*Workload` classes back to `*Test` to match codebase convention (50+ existing classes). See a892135.

File: workloads/ops/logical_reduce.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    def __init__(self, shape: tuple, dtype: torch.dtype):
+        self.shape = shape
+        self.dtype = dtype
+
+    def gen_inputs(self) -> tuple[torch.Tensor]:
+        return (_make_logical_input(self.shape, self.dtype),)
+
+
+class CountNonzeroWorkload(WorkloadBase):
+    """Workload definition for CountNonzeroFwdOp.
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply workloads/ops/logical_reduce.py]: Adopted. Renamed all `*Workload` classes back to `*Test` to match codebase convention (50+ existing classes). See a892135.

File: workloads/ops/argreduce.py [hunk 9]
```diff
@@ -15,13 +15,14 @@ def gen_inputs(self) -> tuple[torch.Tensor]:
...
-    def gen_inputs(self) -> tuple[torch.Tensor]:
-        x = torch.randn(*self.shape, dtype=self.dtype, device="cuda")
-        return (x,)
+class ArgminWorkload(_RandnWorkload):
+    """Workload definition for ArgminFwdOp."""
+
+
+# Backward-compatible aliases for existing consumers
+ArgmaxTest = ArgmaxWorkload
+ArgminTest = ArgminWorkload
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply workloads/ops/argreduce.py]: Adopted. Renamed all `*Workload` classes back to `*Test` to match codebase convention (50+ existing classes). See a892135.

**Review by Ibuki-wind — COMMENTED** [review]
Trust-boundary verdict: clean. This PR stays within the test-owned `tests/` + `workloads/` layer and keeps correctness logic out of workloads.

The workload extraction itself is directionally correct and removes duplicated `gen_inputs()` implementations.

One follow-up: keep the workload-import aliasing consistent with the existing repo pattern so workload classes and test helpers remain unambiguous for future refactors and agent edits.

File: tests/ops/test_reduce.py [hunk 10]
```diff
@@ -131,25 +125,27 @@ def ref_program(self, x: torch.Tensor) -> torch.Tensor:
             return x_f32.amax(dim=-1).to(x.dtype)
         elif self.op_kind == "amin":
             return x_f32.amin(dim=-1).to(x.dtype)
-        elif self.op_kind == "prod":
-            return x_f32.prod(dim=-1).to(x.dtype)
         raise ValueError(f"Unknown op_kind: {self.op_kind}")
 
 
-class WelfordTest(TestBase):
+class ProdTest(ProdTest, TestBase):
```
Comment (Ibuki-wind) [inline tests/ops/test_reduce.py]: `ProdTest` here shadows the imported workload class of the same name, which breaks the repo's usual `_...Workload` alias pattern and makes future refactors/introspection ambiguous -> import the workload as an underscored alias (for example `_ProdTestWorkload`) and inherit from that alias instead.

File: workloads/ops/reduce.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply workloads/ops/reduce.py]: Adopted. Aliased all workload imports with underscore prefix (`_SumTest`, `_ProdTest`, `_StdTest`). See 01a05dd.

File: tests/ops/test_reduce.py (hunk 10 above)
Comment (lcy-seso) [reply tests/ops/test_reduce.py]: Adopted. Aliased all workload imports with underscore prefix (`_SumTest`, `_ProdTest`, `_StdTest`). See 01a05dd.

File: tests/ops/test_reduce.py [hunk 11]
```diff
@@ -131,25 +125,27 @@ def ref_program(self, x: torch.Tensor) -> torch.Tensor:
...
+    """Parameterized test helper for prod op (uses small-range inputs)."""
+
+    def __init__(self, m: int, n: int, dtype: torch.dtype):
+        super().__init__((m, n), dtype)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_reduce.py]: Addressed in 01a05dd.

Evidence:
- Import is aliased as `ProdTest as _ProdTest` (line 12 of tests/ops/test_reduce.py)
- Class definition uses the aliased name: `class ProdTest(_ProdTest, TestBase)` (line 135)
- No F811 redefinition — ruff check passes clean

Current HEAD: d497169

File: tests/ops/test_vector_norm.py [hunk 12]
```diff
@@ -86,27 +87,20 @@ class VectorNorm1DFixture(FixtureBase):
 
 
 # ---------------------------------------------------------------------------
-# TestBase helpers
+# TestBase helpers — inherit gen_inputs() from workload classes
 # ---------------------------------------------------------------------------
 
 
 # Map op_kind to the ord parameter for torch.linalg.vector_norm
 _ORD_MAP = {"l1": 1, "l2": 2, "inf": float("inf")}
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_vector_norm.py:96]: Addressed in d497169.

Evidence:
- Added missing blank line before `class VectorNormTest` (line 98 of tests/ops/test_vector_norm.py)
- E302 resolved — ruff check passes clean

Current HEAD: d497169

File: tests/ops/test_reduce.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_reduce.py]: Already addressed in 01a05dd. The import is now `ProdTest as _ProdTest` (line 12), so the local `ProdTest` class (line 135) no longer shadows the import. No F811.

File: tests/ops/test_vector_norm.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_vector_norm.py:96]: Already addressed in d497169. Two blank lines now precede `VectorNormTest` (line 98).

**Review by Ibuki-wind — APPROVED** [review]
Approved at current head d497169. The prior workload-import alias collision in tests/ops/test_reduce.py is resolved, the naming now matches the repo pattern, and I do not have remaining review findings on this PR.
